enterprise/server/hit_tracker_client/hit_tracker_client.go

```diff
@@ -2,57 +2,298 @@ package hit_tracker_client
+	"github.com/buildbuddy-io/buildbuddy/server/util/grpc_client"
+	"github.com/buildbuddy-io/buildbuddy/server/util/log"
+	"github.com/buildbuddy-io/buildbuddy/server/util/status"
+	"github.com/prometheus/client_golang/prometheus"
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/types/known/durationpb"
 
+	hitpb "github.com/buildbuddy-io/buildbuddy/proto/hit_tracker"
 	repb "github.com/buildbuddy-io/buildbuddy/proto/remote_execution"
+	rspb "github.com/buildbuddy-io/buildbuddy/proto/resource"
+	gstatus "google.golang.org/grpc/status"
 )
 
+const (
+	initialSliceCapacity = 4_096
+)
+
+var (
+	remoteHitTrackerTarget       = flag.String("cache.remote_hit_tracker.target", "", "The gRPC target of the remote cache-hit-tracking service.")
+	remoteHitTrackerTimeout      = flag.Duration("cache.remote_hit_tracker.rpc_timeout", 5*time.Second, "The timout to use for gRPC requests to the remote cache-hit-tracking service.")
+	remoteHitTrackerPollInterval = flag.Duration("cache.remote_hit_tracker.update_interval", 250*time.Millisecond, "The time interval to wait between sending remote cache-hit-tracking RPCs.")
+	maxPendingHitsPerGroup       = flag.Int("cache.remote_hit_tracker.max_pending_hits_per_group", 100_000, "The maximum number of pending cache-hit updates to store in memory for a given group.")
+	remoteHitTrackerWorkers      = flag.Int("cache.remote_hit_tracker.workers", 1, "The number of workers to use to send asynchronous remote cache-hit-tracking RPCs.")
+)
+
+func Register(env *real_environment.RealEnv) error {
+	if *remoteHitTrackerTarget == "" {
+		return nil
+	}
+
+	conn, err := grpc_client.DialInternal(env, *remoteHitTrackerTarget)
+	if err != nil {
+		return err
+	}
+	env.SetHitTrackerFactory(newHitTrackerClient(env, conn))
+	return nil
+}
+
+func newHitTrackerClient(env *real_environment.RealEnv, conn grpc.ClientConnInterface) *HitTrackerFactory {
+	factory := HitTrackerFactory{
+		authenticator:          env.GetAuthenticator(),
+		ticker:                 env.GetClock().NewTicker(*remoteHitTrackerPollInterval).Chan(),
+		quit:                   make(chan struct{}, 1),
+		stopped:                atomic.Bool{},
+		maxPendingHitsPerGroup: *maxPendingHitsPerGroup,
+		hitsByGroup:            map[groupID]*cacheHits{},
+		rpcTimeout:             *remoteHitTrackerTimeout,
+		client:                 hitpb.NewHitTrackerServiceClient(conn),
+	}
+	for i := 0; i < *remoteHitTrackerWorkers; i++ {
+		go factory.runWorker()
+	}
+	env.GetHealthChecker().RegisterShutdownFunction(factory.shutdown)
+	return &factory
+}
+
+type groupID string
+type cacheHits struct {
+	authHeaders map[string][]string
+	hits        []*hitpb.CacheHit
+}
+
 type HitTrackerFactory struct {
-	env environment.Env
+	authenticator interfaces.Authenticator
+
+	ticker  <-chan time.Time
+	quit    chan struct{}
+	stopped atomic.Bool
+
+	mu                     sync.Mutex
+	maxPendingHitsPerGroup int
+	hitsByGroup            map[groupID]*cacheHits
+
+	rpcTimeout time.Duration
```

Comment:
is this used? If not I think we can remove it and the flag.

I'm generally suspicious of client-side RPC timeouts like this because it means you're going to drop stuff when your service gets overloaded. That may be what you want, but often there are better ways to do this (like if the channel was full you could drop there instead)

Reply:
Whoops, unused, removed.